web-animations/timing-model/animations/updating-the-finished-state.html

```diff
@@ -0,0 +1,131 @@
+<!DOCTYPE html>
+<meta charset=utf-8>
+<title>Tests for updating the finished state of an animation</title>
+<link rel="help" href="https://w3c.github.io/web-animations/#updating-the-finished-state">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="../../testcommon.js"></script>
+<body>
+<script>
+'use strict';
+
+function validateFinishEvent(t, event, animation, expectedEventTime) {
+  t.step(function() {
+    assert_equals(event.target, animation, "Animation should be event target");
+    assert_times_equal(event.currentTime, expectedEventTime,
+        "Event currentTime should be the given expected value");
+    assert_times_equal(event.timelineTime, animation.startTime + expectedEventTime,
+        "Event timelineTime should be the animation's start time plus the given expected value");
+  });
+}
+
+function awaitFinishEventAndPromise(t, animation, expectedEventTime) {
+  var eventPromise = new Promise(function(resolve) {
+    animation.onfinish = function(event) {
+      validateFinishEvent(t, event, animation, expectedEventTime);
+      resolve();
+    };
+  });
+
+  var finishedPromise = animation.finished.then(function(target) {
+    assert_equals(target, animation);
+  });
+
+  var timeout = waitForAnimationFrames(3).then(function() {
+    return Promise.reject("Expected finish notifications did not occur");
+  });
+
+  return Promise.race([timeout,Promise.all([eventPromise, finishedPromise])]);
+}
+
+promise_test(function(t) {
+  var div = createDiv(t);
+  var animation = div.animate(null, 1);
+  animation.onfinish = t.unreached_func("Seeking to finish should not fire finish event");
+  animation.finished.then(t.unreached_func("Seeking to finish should not resolve finished promise"));
+  animation.currentTime = 1;
+  animation.currentTime = 0;
+  animation.pause();
+  return waitForAnimationFrames(3);
+}, "Finish notification steps don't run when the animation seeks to finish and then seeks back again");
+
+promise_test(function(t) {
+  var div = createDiv(t);
+  var animation = div.animate(null, 1);
+  animation.onfinish = t.unreached_func("Seeking past finish should not fire finish event");
+  animation.finished.then(t.unreached_func("Seeking past finish should not resolve finished promise"));
+  animation.currentTime = 10;
+  animation.currentTime = 0;
+  animation.pause();
+  return waitForAnimationFrames(3);
+}, "Finish notification steps don't run when the animation seeks past finish and then seeks back again");
+
+promise_test(function(t) {
+  var div = createDiv(t);
+  var animation = div.animate(null, 1);
+  animation.finish();
+  animation.currentTime = 0;
+  animation.pause();
+  var expectedEventTime = 1;
```

Comment:
Thanks for making all these changes, and sorry to take a while to get to this.

I'm not sure that this is right. We called `finish()` so the `timelineTime` for then event should be `document.timeline.currentTime`, right?.

However, we then called `animation.currentTime = 0` which should make  `animation.startTime = document.timeline.currentTime`. Since we have, &ldquo;Set animation’s start time to the result of evaluating timeline time - (seek time / playback rate)&rdquo;, i.e. `start time = document.timeline.currentTime - 0`.

So if we pass `expectedEventTime` as 1, when we call `validateFinishEvent` we'll end up running:

```
assert_times_equal(event.timelineTime, animation.startTime + expectedEventTime)
= assert_times_equal(document.timeline.currentTime, document.timeline.currentTime + 1)
```

which will fail. Am I missing something?